src/value.rs

```diff
@@ -90,6 +102,65 @@ impl AllocatedValue {
     }
 }
 
+impl SignedInteger {
+    pub fn zero() -> Self {
+        SignedInteger::Positive(0)
+    }
+
+    pub fn to_u64(&self) -> Option<u64> {
+        match self {
+            SignedInteger::Positive(x) => Some(*x),
+            SignedInteger::Negative(_) => None,
+        }
+    }
+
+    fn to_i128(&self) -> i128 {
+        match self {
+            SignedInteger::Positive(x) => (*x).into(),
+            SignedInteger::Negative(x) => -(*x as i128),
+        }
+    }
+
+    fn from_i128(x: i128) -> Self {
+        if i128::is_negative(x) {
+            SignedInteger::Negative(-x as u64)
```

Comment:
I'd try to stay safe and write this with explicit order of operations to avoid weird overflows/truncations:

```suggestion
            SignedInteger::Negative((-x) as u64)
```